File: mpmt1.pas
//
//  mpmt1.pas: A stupid simple example of Pascal
//
// License:
//   Apache License, Version 2.0
// History:
//   * 2024/05/20 v0.1 Initial version
// Author:
//   Masanori Itoh <masanori.itoh@gmail.com>
// Usage:
//-  Install FreePascal
//-  $ fpc -ompmt1pas mpmt1.pas
//-  $ ./mpmt1pas  NUM_CONTEXT (number of threads) DURATION (in sec.)
// TODO:
//   * Use argparse
//   * Use (at least) mili-second precision expiration check
//
program mpmt1;

uses cthreads, sysutils, dateutils;

function busy_worker(duration : pointer) : IntPtr;
var
   start_ts : longint;
   current_ts: longint;
begin
   // writeLn('thread ', longint(p), ' started.');
   //DateTimeToUnix(Now) returns timestamp in second
   start_ts := DateTimeToUnix(Now);
   while true do
      begin
         current_ts := DateTimeToUnix(Now);
         if (current_ts - start_ts) > ptruint(duration) then
            begin
               writeln('Expired.');
               exit(0);
            end;
      end;
   //never reach here
   exit(0);
end;

var
   i           : longint;
   duration    : ptruint;
   num_context : longint;
   pidarray    : array of TThreadID;
begin
   duration := 10;
   num_context := 3;
   if paramcount() >= 2 then
      duration := StrToInt(paramStr(2));
   if paramcount() >= 1 then
      num_context := StrToInt(paramStr(1));

   writeln('NUM_CONTEXT: ', num_context, ' DURATION: ', duration);

   //for i:=1 to paramCount() do
   //   Writeln('Opt: ', i, ' = ', StrToInt(paramStr(i)));

   setLength(pidarray, num_context);
   for i:=1 to num_context do
   begin;
      writeln('Creating Thread: ', i);
      pidarray[i] := BeginThread(@busy_worker, pointer(duration));
   end;
   for i:=1 to num_context do
   begin;
      WaitForThreadTerminate(pidarray[i], (duration + 1) * 1000);
      writeln('Thread: ', i, ' returned.');
   end;
end.

File: mpmt1.nim
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# mpmt1.nim: Nim version of mpmit1
#
# License:
#   Apache License, Version 2.0
# History:
#   * 2026/01/28 v0.1 Initial version
# Author:
#   Masanori Itoh <masanori.itoh@gmail.com>
# TODO:
#   * Try malebolgia
#
import std/os
import std/times
import std/threadpool
import std/strutils

proc busyWorker(identity: int, duration: int) {. gcsafe .} =
    echo "busyWorker called. ", identity
    let start = getTime()
    let start_ts = start.toUnix() * 1_000_000_000 + start.nanosecond 
    let end_ts = start_ts + duration * 1_000_000_000

    while true:
        let now = getTime()
        let now_ts = now.toUnix() * 1_000_000_000 + now.nanosecond
        if now_ts >= end_ts:
            echo "Expired."
            break

when isMainModule:
    var num_context = 4
    var duration = 5

    let args = commandLineParams()

    if args.len >= 1:
        num_context = parseInt(args[0])
    if args.len >= 2:
        duration = parseInt(args[1])

    echo "num_context: ", num_context, " duration: ", duration

    for i in 0..num_context - 1:
        spawn busyWorker(i, duration)
    sync()
